src/UseCase/ProvideBannerValues.php

```diff
@@ -0,0 +1,39 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace WMDE\BannerServer\UseCase;
+
+/**
+ * @license GNU GPL v2+
+ */
+class ProvideBannerValues {
+
+	private $impressionCount = 0;
+	private $bannerImpressionCount = 0;
+	private $bucketName;
+	private $campaignName;
+
+	public function __construct( int $impressionCount, int $bannerImpressionCount, ?string $bucketName, ?string $campaignName ) {
```

Comment:
I would not make the strings optional and use empty strings instead. But that's a matter of taste we can discuss.

Reply:
Why do you prefer empty strings over null? I would think that null is semantically more appropriate and I fail to see a real downside to it. The only thing I could come up with is that it makes it so that the variable potentially becomes incompatible with string methods and has the potential to introduce checks for null values, though I am not sure how relevant / realistic that fear really is.

Reply:
Why do you prefer empty strings over null? I would think that null is semantically more appropriate and I fail to see a real downside to it. The only thing I could come up with is that it makes it so that the variable potentially becomes incompatible with string methods and has the potential to introduce checks for null values, though I am not sure how relevant / realistic that fear really is.

Reply:
Why do you prefer empty strings over null? I would think that null is semantically more appropriate and I fail to see a real downside to it. The only thing I could come up with is that it makes it so that the variable potentially becomes incompatible with string methods and has the potential to introduce checks for null values, though I am not sure how relevant / realistic that fear really is.

Reply:
Why do you prefer empty strings over null? I would think that null is semantically more appropriate and I fail to see a real downside to it. The only thing I could come up with is that it makes it so that the variable potentially becomes incompatible with string methods and has the potential to introduce checks for null values, though I am not sure how relevant / realistic that fear really is.

Reply:
Why do you prefer empty strings over null? I would think that null is semantically more appropriate and I fail to see a real downside to it. The only thing I could come up with is that it makes it so that the variable potentially becomes incompatible with string methods and has the potential to introduce checks for null values, though I am not sure how relevant / realistic that fear really is.

Reply:
Why do you prefer empty strings over null? I would think that null is semantically more appropriate and I fail to see a real downside to it. The only thing I could come up with is that it makes it so that the variable potentially becomes incompatible with string methods and has the potential to introduce checks for null values, though I am not sure how relevant / realistic that fear really is.

Reply:
Why do you prefer empty strings over null? I would think that null is semantically more appropriate and I fail to see a real downside to it. The only thing I could come up with is that it makes it so that the variable potentially becomes incompatible with string methods and has the potential to introduce checks for null values, though I am not sure how relevant / realistic that fear really is.